google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -445,43 +521,161 @@ void extendDeadlines() {
       entry.getValue().forget();
       if (totalExpiration.isAfter(now)) {
         int sec = Math.max(1, (int) now.until(totalExpiration, ChronoUnit.SECONDS));
-        modacks.add(new PendingModifyAckDeadline(sec, ackId));
+        ModackRequestData modackRequestData =
+            modackWithMessageFutureByExtensionTimeMap.computeIfAbsent(
+                sec, extensionSeconds -> new ModackRequestData(extensionSeconds));
+        modackRequestData.addAckIdMessageFuture(entry.getValue().getAckRequestData());
+        numAckIdToSend++;
       }
     }
-    logger.log(Level.FINER, "Sending {0} modacks", modack.ackIds.size() + modacks.size());
-    modacks.add(modack);
 
-    List<String> acksToSend = Collections.emptyList();
-    ackProcessor.sendAckOperations(acksToSend, modacks);
+    if (numAckIdToSend > 0) {
+      logger.log(Level.FINER, "Sending {0} modacks", numAckIdToSend);
+      ackProcessor.sendModackOperations(
+          new ArrayList<ModackRequestData>(modackWithMessageFutureByExtensionTimeMap.values()));
+    }
   }
 
   @InternalApi
-  void processOutstandingAckOperations() {
-    List<PendingModifyAckDeadline> modifyAckDeadlinesToSend = new ArrayList<>();
+  void processOutstandingOperations() {
+    List<AckRequestData> ackRequestDataList = new ArrayList<AckRequestData>();
+    pendingAcks.drainTo(ackRequestDataList);
+    logger.log(Level.FINER, "Sending {0} acks", ackRequestDataList.size());
+
+    ackProcessor.sendAckOperations(ackRequestDataList);
```

Comment:
We used to send the ack operations together to the StreamingSubscriber connection, and the ModAcks would be processed before the Acks.

I doubt this will change anything, but just to keep this the same, can we send the modacks before the acks.

Reply:
Makes sense, will change the order